shared/src/sources/balancer/info_fetching.rs

```diff
@@ -83,10 +95,310 @@ impl PoolInfoFetching for PoolInfoFetcher {
             .map(Bfp::from_wei)
             .collect();
         Ok(WeightedPoolInfo {
-            pool_id,
-            tokens,
+            common: CommonPoolInfo {
+                pool_id,
+                tokens,
+                scaling_exponents,
+            },
             weights,
-            scaling_exponents,
         })
     }
+
+    async fn get_stable_pool_data(&self, pool_address: H160) -> Result<StablePoolInfo> {
+        let mut batch = CallBatch::new(self.web3.transport());
+        let pool_contract = BalancerV2StablePool::at(&self.web3, pool_address);
+        let pool_id = H256::from(pool_contract.methods().get_pool_id().call().await?.0);
+
+        // token_data and weight calls can be batched
+        let token_data = self
+            .vault
+            .methods()
+            .get_pool_tokens(Bytes(pool_id.0))
+            .batch_call(&mut batch);
+        let amplification_parameter = pool_contract
+            .methods()
+            .get_amplification_parameter()
+            .batch_call(&mut batch);
+        batch.execute_all(MAX_BATCH_SIZE).await;
+
+        let tokens = token_data.await?.0;
+
+        let token_decimals = self.token_info_fetcher.get_token_infos(&tokens).await;
+        let ordered_decimals = tokens
+            .iter()
+            .map(|token| token_decimals.get(token).and_then(|t| t.decimals))
+            .collect::<Option<Vec<_>>>()
+            .ok_or_else(|| anyhow!("all token decimals required to build scaling factors"))?;
+        // Note that balancer does not support tokens with more than 18 decimals
+        // https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce70f7663e0ac94b25ed60cb86faaa8199fd9e13/pkg/pool-utils/contracts/BasePool.sol#L497-L508
+        let scaling_exponents = ordered_decimals
+            .iter()
+            .map(|decimals| 18u8.checked_sub(*decimals))
+            .collect::<Option<Vec<_>>>()
+            .ok_or_else(|| anyhow!("token with more than 18 decimals"))?;
```

Comment:
nit: This code is duplicated in both `get_*_pool_data` methods and could be in its own method.